main/src/ca/uwaterloo/flix/language/phase/Eraser.scala

```diff
@@ -281,4 +337,67 @@ object Eraser {
 
   private def box(@unused tpe: SimpleType): SimpleType = SimpleType.Object
 
+  private final class SharedContext {
+
+    /**
+      * `(Option, List(Int32)) -> Option$42` means that `Option` is specialized wrt. `Int32` under the name `Option$42`.
+      */
+    private val enumSpecializations: ConcurrentHashMap[(Symbol.EnumSym, List[SimpleType]), Symbol.EnumSym] =
+      new ConcurrentHashMap()
+
+    /**
+      * `(MutList, List(Int32)) -> MutList$42` means that `MutList` is specialized wrt. `Int32` under the name `MutList$42`.
+      */
+    private val structSpecializations: ConcurrentHashMap[(Symbol.StructSym, List[SimpleType]), Symbol.StructSym] =
+      new ConcurrentHashMap()
+
+    /** Returns the specialized version of `sym` according to `targs`, creating a new symbol if not done already. */
+    def getSpecializedEnumName(sym: Symbol.EnumSym, targs: List[SimpleType])(implicit flix: Flix): Symbol.EnumSym = {
+      targs match {
+        case Nil =>
+          // No specialization required.
+          enumSpecializations.computeIfAbsent((sym, targs), _ => sym)
+        case _ =>
+          // Do specialization.
+          enumSpecializations.computeIfAbsent((sym, targs), _ => Symbol.freshEnumSym(sym))
+      }
+    }
+
+    /**
+      * Returns all enum specializations `(sym, targs, newSym)` where `sym` should be
+      * specialized wrt. `targs` under the name `newSym`.
+      */
+    def getEnumSpecializations: List[(Symbol.EnumSym, List[SimpleType], Symbol.EnumSym)] =
+      toList(enumSpecializations)
+
+    /** Returns the specialized version of `sym` according to `targs`, creating a new symbol if not done already. */
+    def getSpecializedStructName(sym: Symbol.StructSym, targs: List[SimpleType])(implicit flix: Flix): Symbol.StructSym = {
+      targs match {
+        case Nil =>
+          // No specialization required.
+          structSpecializations.computeIfAbsent((sym, targs), _ => sym)
+        case _ =>
+          // Do specialization.
+          structSpecializations.computeIfAbsent((sym, targs), _ => Symbol.freshStructSym(sym))
+      }
+    }
+
+    /**
+      * Returns all struct specializations `(sym, targs, newSym)` where `sym` should be
+      * specialized wrt. `targs` under the name `newSym`.
+      */
+    def getStructSpecializations: List[(Symbol.StructSym, List[SimpleType], Symbol.StructSym)] =
+      toList(structSpecializations)
+
+    /** Returns the entries of `m`. */
+    private def toList[A, B](m: ConcurrentHashMap[(A, B), A]): List[(A, B, A)] = {
+      val res = mutable.ListBuffer.empty[(A, B, A)]
+      m.forEach(new BiConsumer[(A, B), A] {
```

Comment:
Is there no Scala magic that allows you to just use a lambda here?

Reply:
Auto complete rewrites it to `(t: (A, B), u: A) => res.append((t._1, t._2, u))` which compiles but the editor thinks its a wrong tuple thing and rewrites it to `((t: (A, B), u: A) => res.append((t._1, t._2, u)))` which still has an editor error. I think the error is just a false trigger here.

So the explicit version has a redundancy squiggle (grey) and the shorthand version has a "Scala compiler will replace this argument list with tuple" error with a red squiggle. Both compile. Which do you prefer?